src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationMatrixNodeProperty.java

```diff
@@ -0,0 +1,176 @@
+            this.grantedPermissions.put(e.getKey(),new HashSet<String>(e.getValue()));
+    }
+
+    public Set<String> getGroups() {
+        return sids;
+    }
+
+    /**
+     * Returns all the (Permission,sid) pairs that are granted, in the multi-map form.
+     *
+     * @return
+     *      read-only. never null.
+     */
+    public Map<Permission,Set<String>> getGrantedPermissions() {
+        return Collections.unmodifiableMap(grantedPermissions);
+    }
+
+    /**
+     * Adds to {@link #grantedPermissions}. Use of this method should be limited
+     * during construction, as this object itself is considered immutable once
+     * populated.
+     */
+    public void add(Permission p, String sid) {
+        Set<String> set = grantedPermissions.get(p);
+        if (set == null)
+            grantedPermissions.put(p, set = new HashSet<String>());
+        set.add(sid);
+        sids.add(sid);
+    }
+
+    @Override
+    public boolean isBlocksInheritance() {
+        return blocksInheritance;
+    }
+
+    @Override
+    public void setBlocksInheritance(boolean blocksInheritance) {
+        this.blocksInheritance = blocksInheritance;
+    }
+
+    private final class AclImpl extends SidACL {
+        @CheckForNull
+        @SuppressFBWarnings(value = "NP_BOOLEAN_RETURN_NULL",
+                justification = "As designed, implements a third state for the ternary logic")
+        protected Boolean hasPermission(Sid sid, Permission p) {
+            if (AuthorizationMatrixNodeProperty.this.hasPermission(toString(sid),p,sid instanceof PrincipalSid)) {
+                return true;
+            }
+            return null;
+        }
+    }
+
+    public SidACL getACL() {
+        return acl;
+    }
+
+    /**
+     * Persist {@link ProjectMatrixAuthorizationStrategy} as a list of IDs that
+     * represent {@link ProjectMatrixAuthorizationStrategy#grantedPermissions}.
+     */
+    public static final class ConverterImpl extends AbstractMatrixPropertyConverter {
+        public boolean canConvert(Class type) {
+            return type == AuthorizationMatrixNodeProperty.class;
+        }
+
+        public AuthorizationProperty createSubject() {
+            return new AuthorizationMatrixNodeProperty();
+        }
+    }
+
+    @Extension
+    public static class DescriptorImpl extends NodePropertyDescriptor implements AuthorizationMatrixPropertyDescriptor<AuthorizationMatrixNodeProperty> {
+
+        @Override
+        public AuthorizationMatrixNodeProperty createProperty() {
+            return new AuthorizationMatrixNodeProperty();
+        }
+
+        @Override
+        public PermissionScope getPermissionScope() {
+            return PermissionScope.COMPUTER;
+        }
+
+        @Override
+        public AuthorizationMatrixNodeProperty newInstance(StaplerRequest req, JSONObject formData) throws FormException {
+            return createNewInstance(req, formData, false);
```

Comment:
Did you check that you cannot return `null` from here?

Reply:
No, but the custom serialization format for permissions makes me doubt that'll work.